File: python/examples/buttons.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2024 Darby Johnston
# All rights reserved.

import feather_tk as ftk

import sys

# Create the context and application.
context = ftk.Context()
app = ftk.App(context, sys.argv, "buttons", "Buttons example")
if app.getExit() != 0:
    sys.exit(app.getExit())

# Create a window.
window = ftk.MainWindow(context, app, "buttons", ftk.Size2I(1280, 960))

# Create a layout.
layout = ftk.VerticalLayout(context)
layout.marginRole = ftk.SizeRole.Margin
scrollWidget = ftk.ScrollWidget(context, ftk.ScrollType.Both)
scrollWidget.border = False
scrollWidget.vStretch = ftk.Stretch.Expanding
scrollWidget.widget = layout
window.setWidget(scrollWidget)
        
# Create push buttons.
groupBox = ftk.GroupBox(context, "Push Buttons", layout)
hLayout = ftk.HorizontalLayout(context, groupBox)
hLayout.spacingRole = ftk.SizeRole.SpacingSmall
pushButton = ftk.PushButton(context, "Click", hLayout)
pushButton.setClickedCallback(lambda: print("Click"))
pushButton = ftk.PushButton(context, "Text", hLayout)
pushButton.icon = "Settings"
pushButton = ftk.PushButton(context, "Disabled", hLayout)
pushButton.enabled = False
        
# Create tool buttons.
groupBox = ftk.GroupBox(context, "Tool Buttons", layout)
hLayout = ftk.HorizontalLayout(context, groupBox)
hLayout.spacingRole = ftk.SizeRole.SpacingSmall
toolButtonGroup = ftk.ButtonGroup(context, ftk.ButtonGroupType.Radio)
toolButtonGroup.setCheckedCallback(lambda index, checked: print("Tool:", index))
hLayout = ftk.HorizontalLayout(context, groupBox)
hLayout.spacingRole = ftk.SizeRole.SpacingSmall
toolIcons = [
    "PlaybackReverse",
    "PlaybackStop",
    "PlaybackForward"
]
hLayout2 = ftk.HorizontalLayout(context, hLayout)
hLayout2.spacingRole = ftk.SizeRole.SpacingTool
for i in range(0, 3):
    toolButton = ftk.ToolButton(context, hLayout2)
    toolButton.checkable = True
    toolButton.checked = 0 == i
    toolButton.icon = toolIcons[i]
    toolButtonGroup.addButton(toolButton)
playStopButton = ftk.ToolButton(context, hLayout)
playStopButton.icon = "PlaybackStop"
playStopButton.checkedIcon = "PlaybackForward"
playStopButton.checkable = True
toolButton = ftk.ToolButton(context, "Text", hLayout)
toolButton = ftk.ToolButton(context, "Text", hLayout)
toolButton.icon = "Settings"
toolButton = ftk.ToolButton(context, "Disabled", hLayout)
toolButton.enabled = False
toolButtonGroup.addButton(toolButton)

# Create check boxes.
groupBox = ftk.GroupBox(context, "Check Boxes", layout)
vLayout = ftk.VerticalLayout(context, groupBox)
vLayout.spacingRole = ftk.SizeRole.SpacingSmall
checkBox = ftk.CheckBox(context, "Check", vLayout)
checkBox.setCheckedCallback(lambda checked: print("Checked:", checked))
checkBox = ftk.CheckBox(context, "Disabled", vLayout)
checkBox.enabled = False

# Create radio buttons.
groupBox = ftk.GroupBox(context, "Radio Buttons", layout)
vLayout = ftk.VerticalLayout(context, groupBox)
vLayout.spacingRole = ftk.SizeRole.SpacingSmall
radioButtonGroup = ftk.ButtonGroup(context, ftk.ButtonGroupType.Radio)
radioButtonGroup.setCheckedCallback(lambda index, checked: print("Radio:", index))
for i in range(0, 3):
    radioButton = ftk.RadioButton(context, "Radio {}".format(i), vLayout)
    radioButton.setCheckedCallback(lambda checked: print("Radio:", checked))
    radioButton.checked = 0 == i
    radioButtonGroup.addButton(radioButton)
radioButton = ftk.RadioButton(context, "Disabled", vLayout)
radioButton.enabled = False

# Show the window and run the application.
window.show()
app.run()

# \bug Need to manually reset the window.
window = None

File: examples/buttons/buttons.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2024-2025 Darby Johnston
// All rights reserved.

#include <feather-tk/ui/App.h>
#include <feather-tk/ui/ButtonGroup.h>
#include <feather-tk/ui/CheckBox.h>
#include <feather-tk/ui/GroupBox.h>
#include <feather-tk/ui/MainWindow.h>
#include <feather-tk/ui/PushButton.h>
#include <feather-tk/ui/RadioButton.h>
#include <feather-tk/ui/RowLayout.h>
#include <feather-tk/ui/ScrollWidget.h>
#include <feather-tk/ui/ToolButton.h>

#include <feather-tk/core/Format.h>

using namespace feather_tk;

FEATHER_TK_MAIN()
{
    try
    {
        // Create the context and application.
        auto context = Context::create();
        auto app = App::create(context, argc, argv, "buttons", "Buttons example");
        if (app->getExit() != 0)
            return app->getExit();

        // Create a window.
        auto window = MainWindow::create(
            context,
            app,
            "buttons",
            Size2I(1280, 960));

        // Create a layout.
        auto layout = VerticalLayout::create(context);
        layout->setMarginRole(SizeRole::Margin);
        auto scrollWidget = ScrollWidget::create(context, ScrollType::Both);
        scrollWidget->setBorder(false);
        scrollWidget->setWidget(layout);
        scrollWidget->setVStretch(Stretch::Expanding);
        window->setWidget(scrollWidget);

        // Create push buttons.
        auto groupBox = GroupBox::create(context, "Push Buttons", layout);
        auto hLayout = HorizontalLayout::create(context, groupBox);
        hLayout->setSpacingRole(SizeRole::SpacingSmall);
        auto pushButton = PushButton::create(context, "Click", hLayout);
        pushButton->setClickedCallback(
            []
            {
                std::cout << "Click" << std::endl;
            });
        pushButton = PushButton::create(context, "Text", hLayout);
        pushButton->setIcon("Settings");
        pushButton = PushButton::create(context, "Disabled", hLayout);
        pushButton->setEnabled(false);

        // Create tool buttons.
        groupBox = GroupBox::create(context, "Tool Buttons", layout);
        hLayout = HorizontalLayout::create(context, groupBox);
        hLayout->setSpacingRole(SizeRole::SpacingSmall);
        auto toolButtonGroup = ButtonGroup::create(context, ButtonGroupType::Radio);
        toolButtonGroup->setCheckedCallback(
            [](int index, bool value)
            {
                std::cout << Format("Tool {0}").arg(index) << std::endl;
            });
        const std::vector<std::string> toolIcons =
        {
            "PlaybackReverse",
            "PlaybackStop",
            "PlaybackForward"
        };
        const std::vector<std::string> toolTooltips =
        {
            "Reverse playback",
            "Stop playback",
            "Forward playback"
        };
        auto hLayout2 = HorizontalLayout::create(context, hLayout);
        hLayout2->setSpacingRole(SizeRole::SpacingTool);
        for (size_t i = 0; i < 3; ++i)
        {
            auto toolButton = ToolButton::create(context, hLayout2);
            toolButton->setCheckable(true);
            toolButton->setChecked(0 == i);
            toolButton->setIcon(toolIcons[i]);
            toolButton->setTooltip(toolTooltips[i]);
            toolButtonGroup->addButton(toolButton);
        }
        auto playStopButton = ToolButton::create(context, hLayout);
        playStopButton->setIcon("PlaybackStop");
        playStopButton->setCheckedIcon("PlaybackForward");
        playStopButton->setCheckable(true);
        auto toolButton = ToolButton::create(context, "Text", hLayout);
        toolButton = ToolButton::create(context, "Text", hLayout);
        toolButton->setIcon("Settings");
        toolButton = ToolButton::create(context, "Disabled", hLayout);
        toolButton->setEnabled(false);
        toolButtonGroup->addButton(toolButton);

        // Create check boxes.
        groupBox = GroupBox::create(context, "Check Boxes", layout);
        auto vLayout = VerticalLayout::create(context, groupBox);
        vLayout->setSpacingRole(SizeRole::SpacingSmall);
        auto checkBox = CheckBox::create(context, "Check", vLayout);
        checkBox->setCheckedCallback(
            [](bool value)
            {
                std::cout << Format("Checked: {0}").arg(value) << std::endl;
            });
        checkBox = CheckBox::create(context, "Disabled", vLayout);
        checkBox->setEnabled(false);

        // Create radio buttons.
        groupBox = GroupBox::create(context, "Radio Buttons", layout);
        vLayout = VerticalLayout::create(context, groupBox);
        vLayout->setSpacingRole(SizeRole::SpacingSmall);
        auto radioButtonGroup = ButtonGroup::create(context, ButtonGroupType::Radio);
        radioButtonGroup->setCheckedCallback(
            [](int index, bool value)
            {
                std::cout << Format("Radio {0}").arg(index) << std::endl;
            });
        for (size_t i = 0; i < 3; ++i)
        {
            auto radioButton = RadioButton::create(context, vLayout);
            radioButton->setText(Format("Radio {0}").arg(i));
            radioButton->setChecked(0 == i);
            radioButtonGroup->addButton(radioButton);
        }
        auto radioButton = RadioButton::create(context, "Disabled", vLayout);
        radioButton->setEnabled(false);

        // Show the window and run the application.
        window->show();
        app->run();
    }
    catch (const std::exception& e)
    {
        std::cout << "ERROR: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

